smartlearner/optimizers/optimizer.py

```diff
@@ -12,12 +12,22 @@ def __init__(self, loss):
         self._update_rules = []
         self._param_modifiers = []
 
+        self.graph_updates = OrderedDict()
+        self._directions = None
+
     def append_update_rule(self, update_rule):
         self._update_rules.append(update_rule)
 
     def append_param_modifier(self, param_modifier):
         self._param_modifiers.append(param_modifier)
 
+    @property
+    def directions(self):
+        if self._directions is None:
```

Comment:
> The main reason of this refactoring is to be able to monitor the gradients as well as the directions (and we should have the same for the loss) so we can print them or compute use them to compute some metrics (e.g. check that the norm of the gradients approaches 0).

Why not just call `self._get_directions()` again? If it changed, it'll track the new one.

> Right now, when adding such task the entire graph is not created yet

If that'd be the case, then this code will also fail since it calls for the gradient from the loss.

> The graph is fully created only once we called `trainer.train`

That's actually false : the optimizer needs a loss at creation.

> The best would be to disallow graph modification once the `Trainer` object is created.

That's unadvisable, since you need to make those modifications, if only to add dropout (which is the first use case that comes to mind).